presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -288,104 +282,34 @@ public PlanWithProperties visitAggregation(AggregationNode node, Context context
             }
 
             if (node.getGroupBy().isEmpty()) {
-                if (decomposable) {
-                    return splitAggregation(node, child, partial -> gatheringExchange(idAllocator.getNextId(), REMOTE, partial));
-                }
-                else {
-                    child = withDerivedProperties(
-                            gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
-                            child.getProperties());
+                child = withDerivedProperties(
+                        gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
+                        child.getProperties());
 
-                    return rebaseAndDeriveProperties(node, child);
-                }
+                return rebaseAndDeriveProperties(node, child);
             }
             else {
                 if (child.getProperties().isStreamPartitionedOn(partitioningRequirement)) {
                     return rebaseAndDeriveProperties(node, child);
                 }
                 else {
                     if (decomposable) {
-                        Function<PlanNode, PlanNode> exchanger = null;
                         if (!child.getProperties().isNodePartitionedOn(partitioningRequirement)) {
-                            exchanger = partial -> partitionedExchange(
-                                    idAllocator.getNextId(),
-                                    REMOTE,
-                                    partial,
-                                    node.getGroupBy(),
-                                    node.getHashSymbol());
+                            child = withDerivedProperties(
+                                    partitionedExchange(idAllocator.getNextId(), REMOTE, child.getNode(), node.getGroupBy(), node.getHashSymbol()),
+                                    child.getProperties());
                         }
-                        return splitAggregation(node, child, exchanger);
                     }
                     else {
                         child = withDerivedProperties(
```

Comment:
i think this clause was missing a
```java
!child.getProperties().isNodePartitionedOn(partitioningRequirement)
```
around it.
we can probably simplify this whole section